### DISCUSSION TITLE
Login to Tibero7 fails with pyodbc 5.x, works with 4.x

### DISCUSSION THREAD

**Discussion by ybjeon01** [mkleehammer/pyodbc#1371] [Q&A] [answered]
**Host OS:** Ubuntu 24.04  
**Container OS:** Ubuntu 22.04  
**Container Python:** Python 3.10  

Running inside a virtual environment.

**PyODBC Version:** 5.1

My `/etc/odbc.ini` configuration is as follows:
```ini
[ODBC]
Trace       = yes
TraceFile   = /tmp/odbc.trace

[Tibero7]
Driver      = Tibero7Driver
Description = Tibero7 ODBC Datasource
SID         = tibero
User        = tibero
Password    = tmax
```

In this environment, executing the following Python code results in an error:

```python
import pyodbc

conn = pyodbc.connect("DSN=Tibero7")
```
Error message:
```
Error: ('0000', '[0000]   (-17001) (SQLDriverConnect)')
```

The error code `-17001` from Tibero indicates: `Login failed: invalid user name or password`.

However, if I uninstall PyODBC 5.1 and install PyODBC 4.0.39, the connection is established without any issues.

Upon reviewing the PyODBC code, it appears that starting from version 5.x, the keys `user` and `password` are transformed to `uid` and `pwd`. I would like to understand if this transformation only applies to the kwargs and connection string, or if it also affects the `odbc.ini` file.

If this transformation does affect the `odbc.ini` file, how can I configure it to send the `user` and `password` keys/values to the driver without transformation?

Thank you for your assistance.

**Comment by gordthompson** [comment]
> Upon reviewing the PyODBC code, it appears that starting from version 5.x, the keys `user` and `password` are transformed to `uid` and `pwd`.

Are you referring to this?

https://github.com/mkleehammer/pyodbc/blob/7a710e760988fdf44494d35e09784960feea2ad5/src/pyodbcmodule.cpp#L479

**Reply by gordthompson** [reply]
Also, an [ODBC trace](https://github.com/mkleehammer/pyodbc/wiki/Troubleshooting-%E2%80%93-Generating-an-ODBC-trace-log) of the failing connection might be helpful.

**Reply by ybjeon01** [reply]
@gordthompson 

Thank you for your response.

Yes, the code is what I am referring to. I am not familiar with reading Python C extension code, so I used the grep command to identify the differences between versions 4.0.39 and 5.1.

When using iODBC, the TraceFile worked, but unfortunately, with unixODBC, the TraceFile was not created, so I was unable to check the trace file.

**Reply by gordthompson** [reply]
Your initial question shows an (otherwise valid) [ODBC] section in odbc.ini. However, with unixODBC that section must be placed in odbcinst.ini, not odbc.ini.

**Reply by ybjeon01** [reply]
Oh, it should have been in odbcinst.ini instead of odbc.ini. I didn't know that. I'll modify the file, and read the trace file. Thank you.

**Reply by gordthompson** [reply]
> I'll … read the trace file.

Please post it as a file attachment here so we can see it, too.

**Reply by ybjeon01** [reply, marked as answer]
I did three tests and read the trace files:

1. Using pyodbc 4.0.39 with encoding setting (success case)
2. Using pyodbc 5.1 without encoding environment variables set for the Tibero ODBC shared library (fail case)
3. Using pyodbc 5.1 with encoding environment variables set for the Tibero ODBC shared library (success case. today, I found a way to use pyodbc 5.1 :D)

The trace files showed that pyodbc 4.0.39 attempts to use SQLConnectW() and, if that fails, it falls back to using SQLConnect(). On the other hand, pyodbc 5.1 only attempts to use SQLConnectW().

here are trace files.

1. Using pyodbc 4.0.39 with encoding setting (success case)
[unixodbctrace-4-with-encoding-setting.log](https://github.com/user-attachments/files/16189152/unixodbctrace-4-has-encoding-setting.log)
![image](https://github.com/user-attachments/assets/763b2ca2-5d75-4399-a43c-76a5aa4e58e9)

2. Using pyodbc 5.1 without encoding environment variables set for the Tibero ODBC shared library (fail case)
[unixodbctrace-5-without-encoding-setting.log](https://github.com/user-attachments/files/16189460/unixodbctrace-5-no-encoding-setting.log)
![image](https://github.com/user-attachments/assets/83f0aae4-9d24-4743-b044-70e41c49ffa8)
yesterday, it did not crash, but today I rebooted my computer and python3 crashes for some reasons.


3. Using pyodbc 5.1 with encoding environment variables set for the Tibero ODBC shared library (success case)
[unixodbctrace-5-with-tbcli-env.log](https://github.com/user-attachments/files/16189462/unixodbctrace-with-tbcli-env.log)
![image](https://github.com/user-attachments/assets/babfa1f5-468a-48f4-ae92-915ac15d5709)
In this case, I found that there is a setting that determines which encoding to use for interpreting connection string. Setting the environment variable `TBCLI_WCHAR_TYPE` to `UCS2` resolves the issue.

**Comment by v-chojas** [comment]
That code only touches the connection string. pyODBC doesn't know about (nor care) the contents of odbc.ini.

**Reply by ybjeon01** [reply]
Thank you for your response.

I think I need to use gdb, or inspect packets with Wireshark, or read the database logs. I only have experience debugging only Python code in my toy project. I guess I need to debug the ODBC shared library and pyodbc as well. If it's alright, could I get some advice on how to do this?
